tensorflow/python/keras/saving/pickle_utils.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Support for Python's Pickle protocol by implementing
+packing/unpacking of Keras Models into objects which are themselves
+picklable.
+"""
+
+import os
+import tarfile
+
+from io import BytesIO
+from uuid import uuid4
+
+from numpy import asarray
+
+from tensorflow.python.keras.saving.save import load_model
+from tensorflow.python.lib.io import file_io
+from tensorflow.python.platform import gfile
+
+
+
+def unpack_model(packed_keras_model):
+  """Reconstruct a Model from the result of pack_model.
+
+  Args:
+      packed_keras_model (np.array): return value from pack_model.
+
+  Returns:
+      tf.keras.Model: a Keras Model instance.
+  """
+  temp_dir = f"ram://{uuid4()}"
+  b = BytesIO(packed_keras_model)
+  with tarfile.open(fileobj=b, mode="r") as archive:
+    for fname in archive.getnames():
+      dest_path = os.path.join(temp_dir, fname)
+      file_io.recursive_create_dir_v2(os.path.dirname(dest_path))
+      with gfile.GFile(dest_path, "wb") as f:
+        f.write(archive.extractfile(fname).read())
+  model = load_model(temp_dir)
+  file_io.delete_recursively_v2(temp_dir)
+  return model
+
+
+def pack_model(model):
+  """Pack a Keras Model into a numpy array for pickling.
+
+  Args:
+      model (tf.keras.Model): a Keras Model instance.
+
+  Returns:
+      tuple: an unpacking function (unpack_model) and it's arguments,
+      as per Python's pickle protocol.
+  """
+  temp_dir = f"ram://{uuid4()}"
+  model.save(temp_dir)
+  b = BytesIO()
+  with tarfile.open(fileobj=b, mode="w") as archive:
+    for root, _, filenames in file_io.walk_v2(temp_dir):
+      for filename in filenames:
+        dest_path = os.path.join(root, filename)
+        with gfile.GFile(dest_path, "rb") as f:
+          info = tarfile.TarInfo(name=os.path.relpath(dest_path, temp_dir))
```

Comment:
I can, but I think I've already drilled down to the root cause as per https://github.com/tensorflow/tensorflow/pull/39609#issuecomment-806052959 and https://github.com/tensorflow/tensorflow/pull/39609#issuecomment-806091177

Reply:
Yes I saw can you add some prints for win in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/lib/io/file_io.py#L818-L824?

Reply:
So something like:
```python
  def _make_full_path(parent, item):
    # Since `os.path.join` discards paths before one that starts with the path
    # separator (https://docs.python.org/3/library/os.path.html#os.path.join),
    # we have to manually handle that case as `/` is a valid character on GCS.
    print(item)
    if item[0] == os.sep:
      print("branch1")
      return "".join([os.path.join(parent, ""), item])
    print("branch1")
    return os.path.join(parent, item)
```
?

I'm not sure exactly what print statements you are looking for. I should have edits from maintainers enabled, so feel free to edit this PR as well.

Reply:
Thanks but I am not a maintainers in this repo and I've not a Win machine. We are inside `walk_v2` there with the CI and few prints we could debug why you receive `\` in filenames on Win.

Reply:
Okay I just pushed some debug statements, can you kick off a test?

Reply:
@bhack  logs are up

Reply:
It looks like `listdir` is returning `\\`. Maybe since a single `\` is a special character it is being escaped to `\\` somewhere, and so then when you strip `\` you still have one `\`?

Reply:
Mhh.. I am not on the desk now but we could check https://github.com/tensorflow/tensorflow/blob/b1d3c4af2ac9470de7eec934c449ef9997f2a2d9/tensorflow/core/platform/ram_file_system.h#L178

Reply:
`std::filesystem::path::preferred_separator` is c++17 and we compile by default with c++14.
You can try to use something like https://stackoverflow.com/questions/12971499/how-to-get-the-file-separator-symbol-in-standard-c-c-or/30261289

Reply:
@bhack  could you clarify what the intention of this comment is?

I feel that this is a bug in TF and should be handled separately from this PR. I am not sure if you are suggesting I implement this fix in this PR / another PR, or if that is just an FYI for me.

Reply:
It could be ok also in a new PR but you need to extend that specific test.

Reply:
Can the TF team handle this like we did with the previous bugs I found via this PR? I am not familiar with the internal workings of the C++ side, and it's a bug not a feature I'm adding.

Reply:
Ok can you isolate in a new ticket just few python lines of a standalone example to reproduce this with `ram://`?

Reply:
Sure thing

Reply:
I opened #48086

Reply:
That thicket was not going to reproduce this issue. Do you need something different?

Reply:
In the meantime here you can test with `tf.saved_model.save`  https://github.com/tensorflow/tensorflow/issues/36453#issuecomment-790220842